pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
+                else:
+                    data[s, a] = self._fhandler.getNode('/', s)
+        return data
 
+    @wraps(Dataset.iterator, assigned=(), updated=())
+    def iterator(self, mode=None, batch_size=None, num_batches=None,
+                 rng=None, data_specs=None, return_tuple=False,
+                 **kwargs):
+        """
+        data_specs : tuple
+            A `(space, source)` tuple. See :ref:`data_specs` for a full
+            description.
+        return_tuple : bool, optional
+            Always return a tuple, even if there is exactly one source
+            of data being returned. Defaults to `False`.
+        """
+        [mode, batch_size, num_batches, rng, data_specs] = self._init_iterator(
+            mode, batch_size, num_batches, rng, data_specs)
+        convert = None
+
+        if data_specs is None:
+            sources = self._get_canonical_sources()
+            spaces = self._get_canonical_spaces()
+            data_specs = (spaces, sources)
+            data = self._get_canonical_data()
+        else:
+            sources = data_specs[1]
+            data = [self.data[s] for s in sources]
+
+        return FiniteDatasetIterator(self,
+                                     mode(data[0].shape[0],
+                                          batch_size,
+                                          num_batches,
+                                          rng),
+                                     data_specs=data_specs,
+                                     return_tuple=return_tuple,
+                                     convert=convert)
+
+    def _get_canonical_sources(self):
+        """
+        Returns a list. The first element will be the source of `X` if present,
+        or of `topo_view` otherwise. The second element will be the source of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        source = []
+        if 'X' in self.sources:
+            source.append('features')
+            self.spaces.set_alias('X', 'features')
+            self.data.set_alias('X', 'features')
+        if 'topo_view' in self.sources:
+            source.append('features')
+            self.spaces.set_alias('topo_view', 'features')
+            self.data.set_alias('topo_view', 'features')
+        if 'y' in self.sources:
+            source.append('targets')
+        return source
+
+    def _get_canonical_spaces(self):
+        """
+        Returns a list. The first element will be the space of `X` if present,
+        or of `topo_view` otherwise. The second element will be the space of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        space = []
+        try:
+            space.append(self.spaces['X'])
+        except KeyError:
+            space.append(self.spaces['topo_view'])
+        try:
+            space.append(self.spaces['y'])
```

Comment:
Returning a list of spaces can make sense, especially if this is what the documentation tells.